Add full smart contract integration testing framework

dezuman opened the pull request:
Only very basic functionality for now, it will be expanded as missing pieces of the stack are implemented. 

Also contains a struct field name tweak for CoData

dezuman committed:
Add very rudimentary WIP smart contract integration testing framework

dezuman committed:
Add ElementAPI for passing debug data to contract execution context

dezuman committed:
Generalize gitignore for testing subrepos

dezuman committed:
Refactor testing harness to be more general, very WIP

dezuman committed:
Add WIP test for the debug data injector ElementAPI

dezuman committed:
Do some cleanup and rustfmt

dezuman committed:
Fix small bug in hypervisor function pop_costack

The issue was that when returning popped data it always wrote back the number of bytes specified as max buffer size, causing an error if the actual popped data size was smaller. It now uses the smallest of buffer size and actual data size.

dezuman committed:
Clean up testing smart contract

dezuman committed:
Add signed int support to DebugData + tests using it

Also some refactoring of duplicated code

dezuman committed:
Rename for descriptiveness

dezuman committed:
Improve doc and struct field names for DebugData ElementAPI

dezuman committed:
Do rustfmt on added elementAPI source file

dezuman committed:
Fix an incorrectly copied comment

dezuman committed:
Add script that runs rustfmt for all files that use it

dezuman committed:
Remove DebugData functionality for signed integers

Regular CoData operations never use signed values, so it doesn't make sense to include it in the testing structure.

dezuman committed:
Rename some CoData fields for clarity

dezuman committed:
Clean up imports for tests + add some comments

Earlz commented on src/element_interfaces/debug_data.rs:
```diff
@@ -0,0 +1,253 @@
+*/
+
+// IDs >= 0x8000_0000 are meant for "internal" usage within the project, e.g. for unit/integration testing.
+pub const DEBUG_DATA_FEATURE: u32 = 0x8000_0001;
```
Change this to 0x4000_0001 to avoid potential conflicts with blockchain provided element IDs which start at 0x8000_0000

dezuman replied:
Small question on that: 

I read this code comment as >= 0x8000_0000 being "private" as in internal to the project, which seems to be a good fit for a testing function. Am I misunderstanding this?

    pub fn call(&self, codata: &mut CoData, element: u32, function: u32) -> Result<ElementResult, NeutronError>{
        let function = function & (!0x8000_0000); //public calls can not set the top bit, which is reserved for private functions
        self.private_call(codata, element, function)
    }

dezuman replied:
Nevermind, I'll change it.

dezuman committed:
Fix CoData stack handling in testing harness + adapt tests

Big commit. Didn't find a good way to split it up in nice chunks without making the partial builds non-functional however, which would make the effort rather pointless. So here we are.

dezuman committed:
Change feature ID of debug data ElemAPI to 0x4000_0001

Apparently >=0x8000_0000 is reserved for chain specific stuff or something like that.

dezuman committed:
Fix missing comments for added ElemAPI functions

Earlz commented:
Looks good! I'll test this out on my machine tomorrow

Earlz commented:
Tried using build.sh for the test compilation, but encountered an error with neturon-star-rt

```
warning: /Users/earlz/neutron-star-rt/Cargo.toml: `panic` setting is ignored for `test` profile
    Updating crates.io index
error: no matching package named `neutron-star` found
location searched: /Users/earlz/neutron-star
required by package `contract-binary v0.1.0 (/Users/earlz/neutron-star/tests/test_debugdata/contract_mirror)`
warning: /Users/earlz/neutron-star-rt/Cargo.toml: `panic` setting is ignored for `test` profile
    Updating crates.io index
error: no matching package named `neutron-star` found
location searched: /Users/earlz/neutron-star
required by package `contract-binary v0.1.0 (/Users/earlz/neutron-star/tests/test_smoke/contract_smoke)`
```

I think the proper fix would be changing the neutron-star-rt dependencies in these tests to link to github rather than local path. This could potentially kill "backwards compatible" testing though, but I also don't think this actually matters too much at this point and is a problem to worry about later

Earlz commented:
Error on my part for failing test builds, merging since this works even though it can use improvements
